File: 322_Coin_Change/322_2.c
#define min(a, b) (a) < (b) ? (a) : (b)
// 完全背包
int coinChange(int* coins, int coinsSize, int amount){
    int dp[amount + 1];
    for (int i = 0; i <= amount; i++) {
        dp[i] = amount + 1;
    }
    dp[0] = 0;
    for (int i = 0; i < coinsSize; i++) {
        for (int j = coins[i]; j <= amount; j++) {
            dp[j] = min(dp[j], dp[j - coins[i]] + 1);
        }
    }
    return dp[amount] == amount + 1 ? -1 : dp[amount];
}

File: 322_Coin_Change/322_2.py
from typing import List

class Solution:  # 超时
    def coinChange(self, coins: List[int], amount: int) -> int:
        res = float("inf")
        coins.sort(reverse=True)

        def dfs(sum_, count):
            nonlocal res
            if sum_ == amount:
                res = min(res, count)
            elif sum_ > amount:
                return
            for coin in coins:
                dfs(sum_ + coin, count + 1)

        dfs(0, 0)
        return -1 if res == float("inf") else res
